### PR TITLE
Granito NPC for Rooted Forest

### PR DESCRIPTION
**Pull request by Rusty-Box**
Granito will be a friendly npc withi Rooted Forest levels. There are 3 variants planned, the small, big and giant Granito.

**Small Granito:**

The default variant, small Granito walk around randomly ocassionally stopping and looking around a bit and some may sit on the ground. They are also semi-smart, i.e. if there is ave ground to stand, they will drop of ledges otherwise they will simply turn around. If there is a wall in their path and the wall is low enough, they will jump over it! Otherwise, again, they just turn around.

And as a little bonus, if they see Tux close by they will also wave at him once and look up at the player if he stand on-top of their head :)

**Big Granito: (sprites incomplete)**

Big Granito are just a bigger version of Granito behaving the same as their small counterpart. The difference being they are a smart object, i.e. they do not walk of ledges! They also cannot jump. They do not wave at Tux.

**Giant Granito: (sprites incomplete)**

Giant Granito are merely meant as a background object. They simply sit around and don't really react to anything happening.

**_Note: All types of Granito, Tux can stand on but also walk past them, making them basically unisolid._**

### "Proof-of-concept" of small Granito behavior video, using scripted Objects:

https://github.com/SuperTux/supertux/assets/20615048/a6e6fa91-f36b-4801-b972-e9ebae7423c3

---

**TODO (Graphics)**

- [x] Finish Small Granito sprites
- [x] Finish Big Granito sprites
- [x] Finish Giant Granito sprites

**TODO (Code - need help)**

- [x] Code Small Granito
- [x] Code Big Granito
- [x] Code Giant Granito
- [x] Merge #2647 to allow for unisolid collision

### CODE REVIEW

**Comment by Rusty-Box** [comment]
@MatusGuy wanted to work on the coding side of things!

**Comment by MatusGuy** [comment]
i have no idea how i managed to not include granito.hpp in the commit

**Comment by FILIPOKonGithup** [comment]
i hope the granito's get added to the game because i like them

**Comment by FILIPOKonGithup** [comment]
bc61b27b-178d-49c2-a5b2-23702b5a1cfa

File: src/badguy/granito.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#ifndef GRANITO_H
+#define GRANITO_H
```
Comment (mrkubax10) [inline src/badguy/granito.hpp:2]: Wrong header guard naming, also missing license info. (Same in granito.cpp)

Comment (Vankata453) [reply src/badguy/granito.hpp:2]: On the coding side, this PR seems to be very WIP so reviewing it now may be pointless.

Comment (mrkubax10) [reply src/badguy/granito.hpp:2]: Yeah I know, but I want to point out the obvious things.

Comment (MatusGuy) [reply src/badguy/granito.hpp:2]: still thanks for warning. Qt creator usually just does default

Comment (Vankata453) [reply src/badguy/granito.hpp:2]: There's also other issues I notice on the surface, but would keep those for when it's ready for review.

**Comment by NickerAban** [comment]
little reference to minecrafk villagers

**Comment by FILIPOKonGithup** [comment]
uhhhh why are you guys not uploading the Giant Granito sprites yet. its just a question

**Comment by MatusGuy** [comment]
they may be uploaded when they're ready

**Comment by FILIPOKonGithup** [comment]
> they may be uploaded when they're ready

ok

**Comment by MatusGuy** [comment]
🙃

**Comment by Rusty-Box** [comment]
**This leaves us with only two more bugs for Small Granito to fix as of now:**
- Granito jumping instead of walking up a slope normally + looping fall animation when walking down slopes
- Granito unable to carry other stationary Granito (.e. standing, sitting)
- Sitting Granito uses walking animation playing a level even though the correct sitting animaton is shown in the editor

**Comment by MatusGuy** [comment]
- [ ] standing on top of two granito at the same time will only trigger one of them

**Comment by Rusty-Box** [comment]
@Vankata453 we may need your help with this ones:
> - Badguys can still push Granito

**Comment by Rusty-Box** [comment]
- [ ] Bug: Granito no longer reacts when Tux stands on them + does not properly carry Tux anymore

It worked before but broke again recently. I think it happen after [65f6f1e](https://github.com/SuperTux/supertux/pull/2635/commits/65f6f1e0bcdebb8fc5d9232248f99ff32fa72e48) though I myself am not 100% sure on that. However it seems likely.

**Comment by Rusty-Box** [comment]
A came across a couple bugs, unfortunatly :/

- [ ] So, while it has been kinda fixed, I stil encounter an very inconsistent bug that has me either fall off (mostly Small Granito) or jitter on top off (mostly Big Granito) Granito. Sometimes it happens sometimes it does not.


https://github.com/SuperTux/supertux/assets/20615048/8954920c-d7ad-4b32-90de-f8e3cccdb7d6


https://github.com/SuperTux/supertux/assets/20615048/7a01314a-2f59-4ded-9f3c-149c5c097003


---

- [ ] Another thing, you can push Granito with carry-able objects an when you do you sometimes gets they play the walking animation despite still standing still.


https://github.com/SuperTux/supertux/assets/20615048/db51a1f8-b28d-4514-ab01-e45ce14bac12


---

- Lastly, though I suspect this one to be more of an issue with how the game handles unisolids, when you get carried by a Granito and they walk on a slope, you fall off.

https://github.com/SuperTux/supertux/assets/20615048/7d36efe6-cc84-4f31-a39d-4a9de1c72f78

**Comment by MatusGuy** [comment]
Omg 😭😭

**Comment by Rusty-Box** [comment]
@MatusGuy Alrighty. Synced the fork so next time you work on this, you can add the Corrupted Giant Granito object types! :)

**Comment by VincentAraneo** [comment]
Hold on. If granito was reused for the game. Does this mean that some unused enemies in the game will return in an update?

**Comment by Rusty-Box** [comment]
> Hold on. If granito was reused for the game. Does this mean that some unused enemies in the game will return in an update?

@VincentAraneo There are a few old concepts of enemies/characters that may find their way into the game at some point, though outside of "Mr. Rocket" from older v0.3.x versions (whose name likely gets changed) there aren't any other instances planned as of yet. Unless you want to count Totem as an "unused" enemy within story mode, that is.

**Comment by FILIPOKonGithup** [comment]
i am still waiting for un-corrupted granitos to get added to the next nightly build

**Comment by MatusGuy** [comment]
Yeah me too...

**Comment by NickerAban** [comment]
a

**Comment by mrkubax10** [comment]
> a

Could you stop sending random, unrelated messages?

**Comment by FILIPOKonGithup** [comment]
ive been waiting too long but the check is still at 32

**Comment by MatusGuy** [comment]
Nah brudda the pressure 😭😭

**Comment by Frostwithasideofsalt** [comment]
> ive been waiting too long but the check is still at 32

I've been waiting too long for you to stop posting comments that contribute nothing to discussions

**Comment by NickerAban** [comment]
> > a
> 
> Could you stop sending random, unrelated messages?

how why? are you upset about something?

**Comment by MatusGuy** [comment]
cuz those messages dont really contribute much

**Comment by VincentAraneo** [comment]
I'm quietly waiting for the non-corrupted granito to come out in the nightly build:

**Comment by Rusty-Box** [comment]
Some ideas for exposing our boys for scripting:
- Detection script, runs a script when spotted by a Small Granito
- Make Granito individual movements/actions scriptable
  - e.g. `Granito.wave();` -> Granito waves/plays their wave action

_Note: To make a scriptable Granito there should be new object type for them called "Scripted" or "Custom". Granito with that type don't do anything by default!_

**Comment by Rusty-Box** [comment]
Giant Granito is still kill-able. Please fix :)

**Comment by FILIPOKonGithup** [comment]
why is it taking to long for them to get added?

**Comment by Frostwithasideofsalt** [comment]
> why is it taking to long for them to get added?

Their never getting added, because nolok killed them all! So sad.

**Comment by MatusGuy** [comment]
They're not getting added because vincent stole industrial assets

**Comment by Rusty-Box** [comment]
> why is it taking to long for them to get added?

They are rather intricite and require a lot of tweaking and fixing since we never had a NPC type object/character in the game. There is also many other places we gotta work on so sometimes stuff falls behind, getting pickup later. Stay patient will you.

**Comment by Rusty-Box** [comment]
@Vankata453 Marty told me awhile ago he may need your help with making Granito scriptable, when you have time that is.

**Comment by Vankata453** [comment]
@Rusty-Box The `SimpleSquirrel` Squirrel API rework I'm currently working on is the perfect opportunity to add that. You can probably skip adding it until then?

**Comment by MatusGuy** [comment]
Agree. Meanwhile, could you review the granito code?

**Comment by Rusty-Box** [comment]
> @Rusty-Box The `SimpleSquirrel` Squirrel API rework I'm currently working on is the perfect opportunity to add that. You can probably skip adding it until then?

Great. So let's get this thing reviewed and work on the scripting after alongside the API thing then

File: src/badguy/granito.cpp [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+    {
+      // sorry im already carrying this guy
+      return WalkingBadguy::collision(other, hit);
+    }
+
+    // sure dude
+    granito->m_carrying = this;
+
+    // yay
+    m_state = STATE_SIT;
```
Comment (Vankata453) [inline src/badguy/granito.cpp:238]: Comments are fine, could follow the convention of uppercase first letter though (and preferably, an end-of-sentence mark).

File: src/badguy/granito.hpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  void restore_original_state();
+
+private:
+  Timer m_walk_interval;
+  State m_state;
+  State m_original_state;
+
+  bool m_has_waved;
+  bool m_stepped_on; /** true if tux was on top of granito last frame */
+  bool m_airborne; /** unfortunately on_ground() sucks */
```
Comment (Vankata453) [inline src/badguy/granito.hpp:85]: Same for the comments.

**Review by Vankata453 — APPROVED** [review]

File: src/badguy/granito.hpp (hunk 1 above)
Comment (MatusGuy) [reply src/badguy/granito.hpp]: How has this never been fixed

Comment (MatusGuy) [reply src/badguy/granito.hpp]: Nvm github.dev fooled me

**Review by mrkubax10 — APPROVED** [review]

**Comment by FILIPOKonGithup** [comment]
why are the granitos only on the linux version of supertux and not the windows version?

**Comment by NickerAban** [comment]
at last the granitos have arrived :'D

**Comment by FILIPOKonGithup** [comment]
idk how many will take for them to add in the windows edition of supertux

**Pull request merged**